odoo/models.py

```diff
@@ -2436,28 +2437,67 @@ def _inherits_join_add(self, current_model, parent_model_name, query):
         )
         return parent_alias
 
+    # TODO: method should be renamed to _related_join_calc
     @api.model
-    def _inherits_join_calc(self, alias, fname, query):
+    def _inherits_join_calc(self, alias, fname, query, allow_related=False):
         """
         Adds missing table select and join clause(s) to ``query`` for reaching
-        the field coming from an '_inherits' parent table (no duplicates).
+        the field coming from an '_inherits' parent table (no duplicates)
+        or from related.
 
         :param alias: name of the initial SQL alias
         :param fname: name of inherited field to reach
         :param query: query object on which the JOIN should be added
+        :param allow_related: whether we make joins for explicit
+                              (i.e. not inherited) related non-stored field
         :return: qualified name of field, to be used in SELECT clause
         """
         # INVARIANT: alias is the SQL alias of model._table in query
         model, field = self, self._fields[fname]
-        while field.inherited:
-            # retrieve the parent model where field is inherited from
-            parent_model = self.env[field.related_field.model_name]
-            parent_fname = field.related[0]
-            # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
-            parent_alias = query.left_join(
-                alias, parent_fname, parent_model._table, 'id', parent_fname,
-            )
-            model, alias, field = parent_model, parent_alias, field.related_field
+        def process_inherited(model, alias, field, check_auto_join=False):
+            while field.inherited:
+                # retrieve the parent model where field is inherited from
+                parent_model = self.env[field.related_field.model_name]
+                parent_fname = field.related[0]
+                if check_auto_join:
+                    join_field = model._fields[parent_fname]
+                    if not join_field.auto_join:
+                        # TODO: user may have access to current model and should not get this access error message
+                        raise AccessError(_("Field %s has %s and can't be used without sudo mode") % (
+                            join_field, 'auto_join=False'
+                        ))
```

Comment:
@rco-odoo What do you think about this TODO? If it's may be useful, shall I implement it in this PR or leave for now?